### PR TITLE
[1731] Add option to change recruitment year for list commands

### PR DESCRIPTION
**Pull request by faissalMT**
### Context
Rollover wasn't yet reflected in MCB.

### Changes proposed in this pull request
Allow the specification of the current recruitment year, or default to the current recruitment year.

### Guidance to review
  
Main concern is test style.  
 [Add option to specify recruitment cycle to mcb commands.](https://trello.com/c/T5oYvmB1/1731-add-option-to-specify-recruitment-cycle-to-mcb-commands)  
  
### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

File: lib/mcb.rb [hunk 1]
```diff
@@ -247,6 +247,9 @@ def iterate_v1_endpoint(url:, endpoint:, **opts)
     def remote_connect_options
       envs = env_to_azure_map.keys.join(', ')
       Proc.new do
+        option :R, 'recruitment_year',
+               "Set the recruitment year, defaults to 2019",
+               argument: :required
```
Comment (defong) [inline lib/mcb.rb]: reads kinda wrong...
argument being required yet defaulting?

Comment (misaka) [reply lib/mcb.rb]: I think it's the `recruitment_year` value that's required.

So this is valid:

```
mcb providers list
mcb providers list -R 2020
```

But not this:

```
mcb providers list -R
```

File: lib/mcb.rb [hunk 2]
```diff
@@ -247,6 +247,9 @@ def iterate_v1_endpoint(url:, endpoint:, **opts)
     def remote_connect_options
       envs = env_to_azure_map.keys.join(', ')
       Proc.new do
+        option :R, 'recruitment_year',
+               "Set the recruitment year, defaults to 2019",
```
Comment (misaka) [inline lib/mcb.rb]: I think the default is the current recruitment cycle, which happens to be 2019 now but will change in the future. Might be better to use that wording.

**Comment by faissalMT** [comment]
[Add option to specify recruitment cycle to mcb commands.](https://trello.com/c/T5oYvmB1/1731-add-option-to-specify-recruitment-cycle-to-mcb-commands)

File: spec/lib/mcb/commands/apiv1/courses/list_spec.rb [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    expect(output).to have_text_table_row('Code',
+                                          'Name',
+                                          'Provider Code',
+                                          'Provider Name')
+    expect(output).to have_text_table_row(course1.course_code, course1.name)
+    expect(output).to have_text_table_row(course2.course_code, course2.name)
+  end
+
+  it 'lists courses for a given recruitment year' do
+    url = "http://localhost:3001/api/v1/2019/courses?recruitment_year=#{next_cycle.year}"
```
Comment (defong) [inline spec/lib/mcb/commands/apiv1/courses/list_spec.rb]: http://localhost:3001/api/v1/`2019`/courses?recruitment_year=`#{next_cycle.year}`

so who wins? `2019` vs `#{next_cycle.year}`

File: spec/lib/mcb/commands/apiv1/providers/list_spec.rb [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    end
+
+    expect(output).to have_text_table_row('Code',
+                                          'Name')
+    expect(output).to have_text_table_row(provider1.provider_code, provider1.provider_name)
+    expect(output).to have_text_table_row(provider2.provider_code, provider2.provider_name)
+  end
+
+  it 'lists providers for a given recruitment year' do
+    url = "http://localhost:3001/api/v1/2019/providers?recruitment_year=#{next_cycle.year}"
```
Comment (defong) [inline spec/lib/mcb/commands/apiv1/providers/list_spec.rb]: same as above who wins? and should it be possible?

File: spec/lib/mcb/commands/providers/list_spec.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        command_output = list(provider1.provider_code)[:stdout]
+
+        expect(command_output).to include(provider1.provider_code)
+        expect(command_output).to include(provider1.provider_name)
+
+        expect(command_output).not_to include(provider2.provider_code)
+        expect(command_output).not_to include(provider2.provider_name)
+      end
+
+      it 'displays all specified providers' do
```
Comment (defong) [inline spec/lib/mcb/commands/providers/list_spec.rb]: why is this one a `it` and this is a `context`

```
context 'when provider is unspecified' do
```

File: lib/mcb.rb [hunk 6]
```diff
@@ -247,6 +247,9 @@ def iterate_v1_endpoint(url:, endpoint:, **opts)
     def remote_connect_options
       envs = env_to_azure_map.keys.join(', ')
       Proc.new do
+        option :R, 'recruitment_year',
```
Comment (misaka) [inline lib/mcb.rb]: Can we use the more standard `recruitment-year` here?

File: lib/mcb.rb [hunk 7]
```diff
@@ -411,6 +414,11 @@ def process_opt_changed_since(opts, url)
         changed_since_param = CGI.escape(changed_since.strftime('%FT%T.%6NZ'))
         new_url.query = "changed_since=#{changed_since_param}"
       end
+
+      if opts.key? :recruitment_year
+        new_url.query = [new_url.query, "recruitment_year=#{opts[:recruitment_year]}"].compact.join('&')
+      end
```
Comment (misaka) [inline lib/mcb.rb]: How about generating a hash and using `to_query` to generate a query from it? Just thinking it'd be a tidier.

```
$ rails c
> { a: 'a', b: 'b' }.to_query
=> "a=a&b=b"
```

**Comment by misaka** [comment]
Loving the new option and the new tests. 😍

File: spec/lib/mcb/commands/providers/list_spec.rb (hunk 5 above)
Comment (faissalMT) [reply spec/lib/mcb/commands/providers/list_spec.rb]: This one is still specific, I'll rephrase it

File: spec/lib/mcb/commands/apiv1/courses/find_spec.rb [hunk 8]
```diff
@@ -10,7 +10,7 @@
     course2 = create(:course, subjects: [subject], site_statuses: [site_status])
 
 
-    url = 'http://localhost:3001/api/v1/2019/courses'
+    url = 'http://localhost:3001/api/v1/courses'
```
Comment (misaka) [inline spec/lib/mcb/commands/apiv1/courses/find_spec.rb]: Just to check, does this simulate the URL that UCAS is currently using? Or is it meant to be the primary way to access the API? To me this looks like it's simulating the URL that UCAS should have stored from the current version of the API.

File: spec/lib/mcb/commands/apiv1/providers/find_spec.rb [hunk 9]
```diff
@@ -9,7 +9,7 @@
   let(:provider2) { create(:provider, sites: [site2], contacts: [contact2]) }
 
   it 'displays the info for the given provider' do
-    url = 'http://localhost:3001/api/v1/2019/providers'
+    url = 'http://localhost:3001/api/v1/providers'
```
Comment (misaka) [inline spec/lib/mcb/commands/apiv1/providers/find_spec.rb]: Same question as for the api v1 courses list.

File: spec/lib/mcb/commands/apiv1/providers/list_spec.rb [hunk 10]
```diff
@@ -0,0 +1,91 @@
+require 'mcb_helper'
+
+describe '"mcb apiv1 providers list"' do
+  let(:current_cycle) { find_or_create(:recruitment_cycle, year: '2019') }
+  let(:next_cycle) { find_or_create(:recruitment_cycle, year: '2020') }
+  let(:provider1) { create(:provider, recruitment_cycle: current_cycle) }
+  let(:provider2) { create(:provider, recruitment_cycle: next_cycle) }
+
+  it 'lists providers' do
+    url = 'http://localhost:3001/api/v1/providers'
```
Comment (misaka) [inline spec/lib/mcb/commands/apiv1/providers/list_spec.rb]: I'm not sure this test makes sense. For API V1 we don't expect to ever get a request without a recruitment cycle, and is documented to only work with the recruitment cycle:

https://github.com/DFE-Digital/manage-courses-backend/blob/master/docs/api.md

I would expect there to be a test to show that if the `-r` option is not provider, only providers from the current cycle are displayed.

File: spec/lib/mcb/commands/providers/list_spec.rb [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        expect(command_output).to include(provider1.provider_code)
+        expect(command_output).to include(provider1.provider_name)
+
+        expect(command_output).to include(provider2.provider_code)
+        expect(command_output).to include(provider2.provider_name)
+      end
+    end
+  end
+
+  context 'when recruitment cycle is specified' do
```
Comment (misaka) [inline spec/lib/mcb/commands/providers/list_spec.rb]: How about a test that when recruitment-cycle is not specified, it defaults to the current one?

File: spec/lib/mcb/commands/courses/list_spec.rb [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        expect(command_output).to include(course1.course_code)
+        expect(command_output).to include(course1.name)
+
+        expect(command_output).to include(course2.course_code)
+        expect(command_output).to include(course2.name)
+      end
+    end
+  end
+
+  context 'when recruitment cycle is specified' do
```
Comment (misaka) [inline spec/lib/mcb/commands/courses/list_spec.rb]: How about a test that when recruitment-cycle is not specified, it defaults to the current one?

File: lib/mcb/commands/providers/list.rb [hunk 13]
```diff
@@ -5,10 +5,10 @@
 run do |opts, args, _cmd|
   MCB.init_rails(opts)
 
-  recruitment_cycle = if opts[:recruitment_year].nil?
+  recruitment_cycle = if !opts.key? :'recruitment-year'
                         RecruitmentCycle.current_recruitment_cycle
                       else
-                        RecruitmentCycle.find_by(year: opts[:recruitment_year])
+                        RecruitmentCycle.find_by(year: opts[:'recruitment-year'])
```
Comment (misaka) [inline lib/mcb/commands/providers/list.rb]: nneeeeewww method!!! :)

File: lib/mcb.rb [hunk 14]
```diff
@@ -113,13 +113,29 @@ def self.generate_apiv2_token(email:, encoding:, secret: nil)
...
 
   def self.each_v1_provider(opts)
+    endpoint = "/api/v1/#{get_recruitment_year(opts)}/providers"
+
     # This method can actually be entirely driven by the args provided, it is auto-configure!
-    iterate_v1_endpoint(**opts)
+    iterate_v1_endpoint(endpoint: endpoint, **opts)
+  end
+
+  def self.get_recruitment_year(opts)
```
Comment (misaka) [inline lib/mcb.rb]: If this method gets called before Rails is initialized it'll fail in potentially a non-obvious way. Can we make it more obvious to the user that Rails is required?

**Pull request merged**